File: examples/python/knapsack_2d_sat.py
#!/usr/bin/env python3
# Copyright 2010-2025 Google LLC
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Solver a 2D rectangle knapsack problem.

This code is adapted from
https://yetanothermathprogrammingconsultant.blogspot.com/2021/10/2d-knapsack-problem.html
"""

import io

from absl import app
from absl import flags
import numpy as np
import pandas as pd

from google.protobuf import text_format

from ortools.sat.python import cp_model


_OUTPUT_PROTO = flags.DEFINE_string(
    "output_proto", "", "Output file to write the cp_model proto to."
)
_PARAMS = flags.DEFINE_string(
    "params",
    "num_search_workers:16,log_search_progress:true,max_time_in_seconds:45",
    "Sat solver parameters.",
)
_MODEL = flags.DEFINE_string(
    "model", "rotation", "'duplicate' or 'rotation' or 'optional'"
)


def build_data() -> tuple[pd.Series, int, int]:
    """Build the data frame."""
    data = """
    item         width    height available    value    color
    k1             20       4       2        338.984   blue
    k2             12      17       6        849.246   orange
    k3             20      12       2        524.022   green
    k4             16       7       9        263.303   red
    k5              3       6       3        113.436   purple
    k6             13       5       3        551.072   brown
    k7              4       7       6         86.166   pink
    k8              6      18       8        755.094   grey
    k9             14       2       7        223.516   olive
    k10             9      11       5        369.560   cyan
    """

    data = pd.read_table(io.StringIO(data), sep=r"\s+")
    print("Input data")
    print(data)

    max_height = 20
    max_width = 30

    print(f"Container max_width:{max_width} max_height:{max_height}")
    print(f"#Items: {len(data.index)}")
    return (data, max_height, max_width)


def solve_with_duplicate_items(
    data: pd.Series, max_height: int, max_width: int
) -> None:
    """solve the problem by building 2 items (rotated or not) for each item."""
    # Derived data (expanded to individual items).
    data_widths = data["width"].to_numpy()
    data_heights = data["height"].to_numpy()
    data_availability = data["available"].to_numpy()
    data_values = data["value"].to_numpy()

    # Non duplicated items data.
    base_item_widths = np.repeat(data_widths, data_availability)
    base_item_heights = np.repeat(data_heights, data_availability)
    base_item_values = np.repeat(data_values, data_availability)
    num_data_items = len(base_item_values)

    # Create rotated items by duplicating.
    item_widths = np.concatenate((base_item_widths, base_item_heights))
    item_heights = np.concatenate((base_item_heights, base_item_widths))
    item_values = np.concatenate((base_item_values, base_item_values))

    num_items = len(item_values)

    # OR-Tools model
    model = cp_model.CpModel()

    # Variables
    x_starts = []
    x_ends = []
    y_starts = []
    y_ends = []
    is_used = []
    x_intervals = []
    y_intervals = []

    for i in range(num_items):
        ## Is the item used?
        is_used.append(model.new_bool_var(f"is_used{i}"))

        ## Item coordinates.
        x_starts.append(model.new_int_var(0, max_width, f"x_start{i}"))
        x_ends.append(model.new_int_var(0, max_width, f"x_end{i}"))
        y_starts.append(model.new_int_var(0, max_height, f"y_start{i}"))
        y_ends.append(model.new_int_var(0, max_height, f"y_end{i}"))

        ## Interval variables.
        x_intervals.append(
            model.new_interval_var(
                x_starts[i],
                item_widths[i] * is_used[i],
                x_ends[i],
                f"x_interval{i}",
            )
        )
        y_intervals.append(
            model.new_interval_var(
                y_starts[i],
                item_heights[i] * is_used[i],
                y_ends[i],
                f"y_interval{i}",
            )
        )

        # Unused boxes are fixed at (0.0).
        model.add(x_starts[i] == 0).only_enforce_if(~is_used[i])
        model.add(y_starts[i] == 0).only_enforce_if(~is_used[i])

    # Constraints.

    ## Only one of non-rotated/rotated pair can be used.
    for i in range(num_data_items):
        model.add(is_used[i] + is_used[i + num_data_items] <= 1)

    ## 2D no overlap.
    model.add_no_overlap_2d(x_intervals, y_intervals)

    ## Objective.
    model.maximize(cp_model.LinearExpr.weighted_sum(is_used, item_values))

    # Output proto to file.
    if _OUTPUT_PROTO.value:
        print(f"Writing proto to {_OUTPUT_PROTO.value}")
        with open(_OUTPUT_PROTO.value, "w") as text_file:
            text_file.write(str(model))

    # Solve model.
    solver = cp_model.CpSolver()
    if _PARAMS.value:
        text_format.Parse(_PARAMS.value, solver.parameters)

    status = solver.solve(model)

    # Report solution.
    if status in (cp_model.OPTIMAL, cp_model.FEASIBLE):
        used = {i for i in range(num_items) if solver.boolean_value(is_used[i])}
        data = pd.DataFrame(
            {
                "x_start": [solver.value(x_starts[i]) for i in used],
                "y_start": [solver.value(y_starts[i]) for i in used],
                "item_width": [item_widths[i] for i in used],
                "item_height": [item_heights[i] for i in used],
                "x_end": [solver.value(x_ends[i]) for i in used],
                "y_end": [solver.value(y_ends[i]) for i in used],
                "item_value": [item_values[i] for i in used],
            }
        )
        print(data)


def solve_with_duplicate_optional_items(
    data: pd.Series, max_height: int, max_width: int
):
    """solve the problem by building 2 optional items (rotated or not) for each item."""
    # Derived data (expanded to individual items).
    data_widths = data["width"].to_numpy()
    data_heights = data["height"].to_numpy()
    data_availability = data["available"].to_numpy()
    data_values = data["value"].to_numpy()

    # Non duplicated items data.
    base_item_widths = np.repeat(data_widths, data_availability)
    base_item_heights = np.repeat(data_heights, data_availability)
    base_item_values = np.repeat(data_values, data_availability)
    num_data_items = len(base_item_values)

    # Create rotated items by duplicating.
    item_widths = np.concatenate((base_item_widths, base_item_heights))
    item_heights = np.concatenate((base_item_heights, base_item_widths))
    item_values = np.concatenate((base_item_values, base_item_values))

    num_items = len(item_values)

    # OR-Tools model
    model = cp_model.CpModel()

    # Variables
    x_starts = []
    y_starts = []
    is_used = []
    x_intervals = []
    y_intervals = []

    for i in range(num_items):
        ## Is the item used?
        is_used.append(model.new_bool_var(f"is_used{i}"))

        ## Item coordinates.
        x_starts.append(
            model.new_int_var(0, max_width - int(item_widths[i]), f"x_start{i}")
        )
        y_starts.append(
            model.new_int_var(0, max_height - int(item_heights[i]), f"y_start{i}")
        )

        ## Interval variables.
        x_intervals.append(
            model.new_optional_fixed_size_interval_var(
                x_starts[i], item_widths[i], is_used[i], f"x_interval{i}"
            )
        )
        y_intervals.append(
            model.new_optional_fixed_size_interval_var(
                y_starts[i], item_heights[i], is_used[i], f"y_interval{i}"
            )
        )
        # Unused boxes are fixed at (0.0).
        model.add(x_starts[i] == 0).only_enforce_if(~is_used[i])
        model.add(y_starts[i] == 0).only_enforce_if(~is_used[i])

    # Constraints.

    ## Only one of non-rotated/rotated pair can be used.
    for i in range(num_data_items):
        model.add(is_used[i] + is_used[i + num_data_items] <= 1)

    ## 2D no overlap.
    model.add_no_overlap_2d(x_intervals, y_intervals)

    ## Objective.
    model.maximize(cp_model.LinearExpr.weighted_sum(is_used, item_values))

    # Output proto to file.
    if _OUTPUT_PROTO.value:
        print(f"Writing proto to {_OUTPUT_PROTO.value}")
        with open(_OUTPUT_PROTO.value, "w") as text_file:
            text_file.write(str(model))

    # solve model.
    solver = cp_model.CpSolver()
    if _PARAMS.value:
        text_format.Parse(_PARAMS.value, solver.parameters)

    status = solver.solve(model)

    # Report solution.
    if status in (cp_model.OPTIMAL, cp_model.FEASIBLE):
        used = {i for i in range(num_items) if solver.boolean_value(is_used[i])}
        data = pd.DataFrame(
            {
                "x_start": [solver.value(x_starts[i]) for i in used],
                "y_start": [solver.value(y_starts[i]) for i in used],
                "item_width": [item_widths[i] for i in used],
                "item_height": [item_heights[i] for i in used],
                "x_end": [solver.value(x_starts[i]) + item_widths[i] for i in used],
                "y_end": [solver.value(y_starts[i]) + item_heights[i] for i in used],
                "item_value": [item_values[i] for i in used],
            }
        )
        print(data)


def solve_with_rotations(data: pd.Series, max_height: int, max_width: int):
    """solve the problem by rotating items."""
    # Derived data (expanded to individual items).
    data_widths = data["width"].to_numpy()
    data_heights = data["height"].to_numpy()
    data_availability = data["available"].to_numpy()
    data_values = data["value"].to_numpy()

    item_widths = np.repeat(data_widths, data_availability)
    item_heights = np.repeat(data_heights, data_availability)
    item_values = np.repeat(data_values, data_availability)

    num_items = len(item_widths)

    # OR-Tools model.
    model = cp_model.CpModel()

    # Coordinate variables for each rectangle.
    x_starts = []
    x_sizes = []
    x_ends = []
    y_starts = []
    y_sizes = []
    y_ends = []
    x_intervals = []
    y_intervals = []

    for i in range(num_items):
        sizes = [0, int(item_widths[i]), int(item_heights[i])]
        # X coordinates.
        x_starts.append(model.new_int_var(0, max_width, f"x_start{i}"))
        x_sizes.append(
            model.new_int_var_from_domain(
                cp_model.Domain.FromValues(sizes), f"x_size{i}"
            )
        )
        x_ends.append(model.new_int_var(0, max_width, f"x_end{i}"))

        # Y coordinates.
        y_starts.append(model.new_int_var(0, max_height, f"y_start{i}"))
        y_sizes.append(
            model.new_int_var_from_domain(
                cp_model.Domain.FromValues(sizes), f"y_size{i}"
            )
        )
        y_ends.append(model.new_int_var(0, max_height, f"y_end{i}"))

        ## Interval variables
        x_intervals.append(
            model.new_interval_var(x_starts[i], x_sizes[i], x_ends[i], f"x_interval{i}")
        )
        y_intervals.append(
            model.new_interval_var(y_starts[i], y_sizes[i], y_ends[i], f"y_interval{i}")
        )

    # is_used[i] == True if and only if item i is selected.
    is_used = []

    # Constraints.

    ## for each item, decide is unselected, no_rotation, rotated.
    for i in range(num_items):
        not_selected = model.new_bool_var(f"not_selected_{i}")
        no_rotation = model.new_bool_var(f"no_rotation_{i}")
        rotated = model.new_bool_var(f"rotated_{i}")

        ### Exactly one state must be chosen.
        model.add_exactly_one(not_selected, no_rotation, rotated)

        ### Define height and width according to the state.
        dim1 = item_widths[i]
        dim2 = item_heights[i]
        # Unused boxes are fixed at (0.0).
        model.add(x_sizes[i] == 0).only_enforce_if(not_selected)
        model.add(y_sizes[i] == 0).only_enforce_if(not_selected)
        model.add(x_starts[i] == 0).only_enforce_if(not_selected)
        model.add(y_starts[i] == 0).only_enforce_if(not_selected)
        # Sizes are fixed by the rotation.
        model.add(x_sizes[i] == dim1).only_enforce_if(no_rotation)
        model.add(y_sizes[i] == dim2).only_enforce_if(no_rotation)
        model.add(x_sizes[i] == dim2).only_enforce_if(rotated)
        model.add(y_sizes[i] == dim1).only_enforce_if(rotated)

        is_used.append(~not_selected)

    ## 2D no overlap.
    model.add_no_overlap_2d(x_intervals, y_intervals)

    # Objective.
    model.maximize(cp_model.LinearExpr.weighted_sum(is_used, item_values))

    # Output proto to file.
    if _OUTPUT_PROTO.value:
        print(f"Writing proto to {_OUTPUT_PROTO.value}")
        with open(_OUTPUT_PROTO.value, "w") as text_file:
            text_file.write(str(model))

    # solve model.
    solver = cp_model.CpSolver()
    if _PARAMS.value:
        text_format.Parse(_PARAMS.value, solver.parameters)

    status = solver.solve(model)

    # Report solution.
    if status in (cp_model.OPTIMAL, cp_model.FEASIBLE):
        used = {i for i in range(num_items) if solver.boolean_value(is_used[i])}
        data = pd.DataFrame(
            {
                "x_start": [solver.value(x_starts[i]) for i in used],
                "y_start": [solver.value(y_starts[i]) for i in used],
                "item_width": [solver.value(x_sizes[i]) for i in used],
                "item_height": [solver.value(y_sizes[i]) for i in used],
                "x_end": [solver.value(x_ends[i]) for i in used],
                "y_end": [solver.value(y_ends[i]) for i in used],
                "item_value": [item_values[i] for i in used],
            }
        )
        print(data)


def main(_):
    """solve the problem with all models."""
    data, max_height, max_width = build_data()
    if _MODEL.value == "duplicate":
        solve_with_duplicate_items(data, max_height, max_width)
    elif _MODEL.value == "optional":
        solve_with_duplicate_optional_items(data, max_height, max_width)
    else:
        solve_with_rotations(data, max_height, max_width)


if __name__ == "__main__":
    app.run(main)

File: examples/cpp/knapsack_2d_sat.cc
// Copyright 2010-2025 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// This file solves a 2D Bin Packing problem as a 2D Knapsack problem.
// It loads the size of the mainrectangle, all available items (rectangles too),
// and tries to fit as many rectangles as possible in the main rectangle.

#include <algorithm>
#include <cstdint>
#include <limits>
#include <string>
#include <vector>

#include "absl/base/log_severity.h"
#include "absl/flags/flag.h"
#include "absl/log/globals.h"
#include "absl/types/span.h"
#include "google/protobuf/text_format.h"
#include "ortools/base/commandlineflags.h"
#include "ortools/base/init_google.h"
#include "ortools/base/logging.h"
#include "ortools/packing/binpacking_2d_parser.h"
#include "ortools/packing/multiple_dimensions_bin_packing.pb.h"
#include "ortools/sat/cp_model.h"
#include "ortools/sat/cp_model_solver.h"

ABSL_FLAG(std::string, input, "", "Input file.");
ABSL_FLAG(int, instance, -1, "Instance number if the file.");
ABSL_FLAG(std::string, params, "", "Sat parameters in text proto format.");

namespace operations_research {
namespace sat {

// Logs the current cost and fills solution_in_ascii_form with a nice ascii
// drawing of the current solution.
void CheckAndPrint2DSolution(
    const CpSolverResponse& response,
    const packing::MultipleDimensionsBinPackingProblem& problem,
    absl::Span<const std::vector<IntervalVar>> interval_by_item_dimension,
    std::string* solution_in_ascii_form) {
  const int num_items = problem.items_size();

  const int64_t objective = response.objective_value();

  const int size_x = problem.box_shape().dimensions(0);
  const int size_y = problem.box_shape().dimensions(1);
  const int64_t box_area = size_x * size_y;
  std::vector<std::vector<int>> values(size_y);
  for (int i = 0; i < size_y; ++i) {
    values[i].resize(size_x, -1);
  }
  int64_t used_area = 0;
  for (int item = 0; item < num_items; ++item) {
    if (!SolutionBooleanValue(
            response, interval_by_item_dimension[item][0].PresenceBoolVar())) {
      continue;
    }
    const int64_t x = SolutionIntegerValue(
        response, interval_by_item_dimension[item][0].StartExpr());
    const int64_t y = SolutionIntegerValue(
        response, interval_by_item_dimension[item][1].StartExpr());
    const int64_t dx = SolutionIntegerValue(
        response, interval_by_item_dimension[item][0].SizeExpr());
    const int64_t dy = SolutionIntegerValue(
        response, interval_by_item_dimension[item][1].SizeExpr());
    used_area += dx * dy;
    for (int i = x; i < x + dx; ++i) {
      for (int j = y; j < y + dy; ++j) {
        if (i >= size_x || j >= size_y) {
          LOG(WARNING) << "Out of shape box: item = " << item << ", x = " << x
                       << ", y = " << y << ", dx = " << dx << ", dy = " << dy;
        } else {
          if (values[i][j] != -1) {
            LOG(WARNING) << "Item " << item << " overlaps with item "
                         << values[i][j];
          }
          values[i][j] = item;
        }
      }
    }
  }

  if (num_items - objective <= 1) {
    LOG(INFO) << "Cost " << objective << ", " << num_items - objective
              << " item selected, area used: " << used_area << "/" << box_area;
  } else {
    LOG(INFO) << "Cost " << objective << ", " << num_items - objective
              << " items selected, area used: " << used_area << "/" << box_area;
  }

  solution_in_ascii_form->clear();
  for (int j = 0; j < size_y; ++j) {
    for (int i = 0; i < size_x; ++i) {
      const int v = values[i][j];
      (*solution_in_ascii_form) += v == -1 ? ' ' : 'A' + v;
    }
    solution_in_ascii_form->append("\n");
  }
}

// Load a 2d bin-packing problem and solve it as a 2d knapsack problem.
// That is fit the max number of object in one box.
void LoadAndSolve(const std::string& file_name, int instance) {
  packing::BinPacking2dParser parser;
  if (!parser.Load2BPFile(file_name, instance)) {
    LOG(FATAL) << "Cannot read instance " << instance << " from file "
               << file_name;
  }
  packing::MultipleDimensionsBinPackingProblem problem = parser.problem();
  LOG(INFO) << "Successfully loaded instance " << instance << " from file "
            << file_name;
  LOG(INFO) << "Instance has " << problem.items_size() << " items";

  const auto box_dimensions = problem.box_shape().dimensions();
  const int num_dimensions = box_dimensions.size();
  const int num_items = problem.items_size();

  CpModelBuilder cp_model;

  // Selects the right shape for each item (plus nil shape if not selected).
  // The nil shape is the first choice.
  std::vector<std::vector<BoolVar>> selected(num_items);
  for (int item = 0; item < num_items; ++item) {
    const int num_shapes = problem.items(item).shapes_size();
    LOG(INFO) << "  - item " << item << " has " << num_shapes << " shapes";
    selected[item].resize(num_shapes + 1);
    for (int shape = 0; shape <= num_shapes; ++shape) {
      selected[item][shape] = cp_model.NewBoolVar();
    }
  }

  // Exactly one shape is selected for each item.
  for (int item = 0; item < num_items; ++item) {
    cp_model.AddEquality(LinearExpr::Sum(selected[item]), 1);
  }

  // Manages positions and sizes for each item.
  std::vector<std::vector<IntervalVar>> interval_by_item_dimension(num_items);
  for (int item = 0; item < num_items; ++item) {
    interval_by_item_dimension[item].resize(num_dimensions);
    const int num_shapes = problem.items(item).shapes_size();
    for (int dim = 0; dim < num_dimensions; ++dim) {
      if (num_shapes == 1) {
        const int64_t dimension = box_dimensions[dim];
        const int64_t size = problem.items(item).shapes(0).dimensions(dim);
        IntVar start = cp_model.NewIntVar({0, dimension - size});
        interval_by_item_dimension[item][dim] =
            cp_model.NewOptionalFixedSizeIntervalVar(start, size,
                                                     selected[item][1]);
      } else {
        const Domain dimension(0, box_dimensions[dim]);
        const IntVar start = cp_model.NewIntVar(dimension);
        const IntVar size = cp_model.NewIntVar(dimension);
        const IntVar end = cp_model.NewIntVar(dimension);
        interval_by_item_dimension[item][dim] =
            cp_model.NewIntervalVar(start, size, end);

        for (int shape = 0; shape <= num_shapes; ++shape) {
          const int64_t item_size_in_dim =
              shape == 0
                  ? 0
                  : problem.items(item).shapes(shape - 1).dimensions(dim);
          cp_model.AddEquality(size, item_size_in_dim)
              .OnlyEnforceIf(selected[item][shape]);
        }
      }
    }
  }

  // Non overlapping.
  if (num_dimensions == 1) {
    LOG(FATAL) << "One dimension is not supported.";
  } else if (num_dimensions == 2) {
    LOG(INFO) << "Box size: " << box_dimensions[0] << "*" << box_dimensions[1];
    NoOverlap2DConstraint no_overlap_2d = cp_model.AddNoOverlap2D();
    for (int item = 0; item < num_items; ++item) {
      no_overlap_2d.AddRectangle(interval_by_item_dimension[item][0],
                                 interval_by_item_dimension[item][1]);
    }
  } else {
    LOG(FATAL) << num_dimensions << " dimensions not supported.";
  }

  // Objective.
  LinearExpr objective;
  for (int item_id = 0; item_id < num_items; ++item_id) {
    objective += selected[item_id][0] * problem.items(item_id).value();
  }
  cp_model.Minimize(objective);

  Model model;
  // Setup parameters.
  SatParameters parameters;
  parameters.set_log_search_progress(true);
  // Parse the --params flag.
  if (!absl::GetFlag(FLAGS_params).empty()) {
    CHECK(google::protobuf::TextFormat::MergeFromString(
        absl::GetFlag(FLAGS_params), &parameters))
        << absl::GetFlag(FLAGS_params);
  }
  model.Add(NewSatParameters(parameters));
  std::string solution_in_ascii_form;
  model.Add(NewFeasibleSolutionObserver([&](const CpSolverResponse& r) {
    if (num_dimensions == 2) {
      CheckAndPrint2DSolution(r, problem, interval_by_item_dimension,
                              &solution_in_ascii_form);
    }
  }));

  const CpSolverResponse response = SolveCpModel(cp_model.Build(), &model);

  if (!solution_in_ascii_form.empty()) {
    LOG(INFO) << solution_in_ascii_form;
  }
}

}  // namespace sat
}  // namespace operations_research

int main(int argc, char** argv) {
  absl::SetStderrThreshold(absl::LogSeverityAtLeast::kInfo);
  InitGoogle(argv[0], &argc, &argv, true);
  if (absl::GetFlag(FLAGS_input).empty()) {
    LOG(FATAL) << "Please supply a data file with --input=";
  }
  if (absl::GetFlag(FLAGS_instance) == -1) {
    LOG(FATAL) << "Please supply a valid instance number with --instance=";
  }

  operations_research::sat::LoadAndSolve(absl::GetFlag(FLAGS_input),
                                         absl::GetFlag(FLAGS_instance));
  return EXIT_SUCCESS;
}
